include/oneapi/dpl/pstl/glue_algorithm_ranges_impl.h

```diff
@@ -16,18 +16,299 @@
 #ifndef _ONEDPL_GLUE_ALGORITHM_RANGES_IMPL_H
 #define _ONEDPL_GLUE_ALGORITHM_RANGES_IMPL_H
 
+#    include <utility>
 #include "execution_defs.h"
-#include "glue_algorithm_defs.h"
 
 #if _ONEDPL_HETERO_BACKEND
 #    include "hetero/algorithm_ranges_impl_hetero.h"
 #    include "hetero/algorithm_impl_hetero.h" //TODO: for __brick_copy
 #endif
 
+#if _ONEDPL_CPP20_RANGES_PRESENT
+#    include <ranges>
+#    include <functional>
+#    include <type_traits>
+#    include <iterator>
+#    include "algorithm_ranges_impl.h"
+#endif
+
 namespace oneapi
 {
 namespace dpl
 {
+
+#if _ONEDPL_CPP20_RANGES_PRESENT
+namespace ranges
+{
+
+// [alg.foreach]
+
+namespace __internal
+{
+
+struct __for_each_fn
+{
+    template<typename _ExecutionPolicy, std::ranges::random_access_range _R, typename _Proj = std::identity,
+             std::indirectly_unary_invocable<std::projected<std::ranges::iterator_t<_R>, _Proj>>  _Fun>
+    requires oneapi::dpl::is_execution_policy_v<std::remove_cvref_t<_ExecutionPolicy>> && std::ranges::sized_range<_R>
+
+    std::ranges::borrowed_iterator_t<_R>
+    operator()(_ExecutionPolicy&& __exec, _R&& __r, _Fun __f, _Proj __proj = {}) const
+    {
+        const auto __dispatch_tag = oneapi::dpl::__ranges::__select_backend(__exec);
+        oneapi::dpl::__internal::__ranges::__pattern_for_each(
+            __dispatch_tag, std::forward<_ExecutionPolicy>(__exec), std::forward<_R>(__r), __f, __proj);
+
+        return {std::ranges::begin(__r) + std::ranges::size(__r)};
+    }
+}; //__for_each_fn
+}  //__internal
+
+inline constexpr __internal::__for_each_fn for_each;
+
+// [alg.transform]
+
+namespace __internal
+{
+struct __transform_fn
+{
+    template<typename _ExecutionPolicy, std::ranges::random_access_range _R,
+             std::ranges::random_access_range _OutRange, std::copy_constructible _F, typename _Proj = std::identity>
+    requires oneapi::dpl::is_execution_policy_v<std::remove_cvref_t<_ExecutionPolicy>> && std::ranges::sized_range<_R>
+        && std::ranges::sized_range<_OutRange> && std::indirectly_writable<std::ranges::iterator_t<_OutRange>,
+                 std::indirect_result_t<_F&, std::projected<std::ranges::iterator_t<_R>, _Proj>>>
+
+    std::ranges::unary_transform_result<std::ranges::borrowed_iterator_t<_R>, std::ranges::borrowed_iterator_t<_OutRange>>
+    operator()(_ExecutionPolicy&& __exec, _R&& __r, _OutRange&& __out_r, _F __op, _Proj __proj = {}) const
+    {
+        const auto __dispatch_tag = oneapi::dpl::__ranges::__select_backend(__exec);
+
+        const auto _size = std::ranges::min(std::ranges::size(__r), std::ranges::size(__out_r));
+
+        oneapi::dpl::__internal::__ranges::__pattern_transform(__dispatch_tag, std::forward<_ExecutionPolicy>(__exec),
+            std::ranges::take_view(__r, _size), std::ranges::take_view(__out_r, _size), __op, __proj);
+
+        return {std::ranges::begin(__r) + _size, std::ranges::begin(__out_r) +  _size};
```

Comment:
This is related to the previous question related to `_size`. 

According to the standard, the returned iterator of the input range iterator must be "last".
"
> A unary_transform_result contains an input iterator equal to last and an output iterator to the element past the last element transformed.

Reply:
For the rage versions with policy we define a little  bit another "size semantic" - see the paper for details.

Reply:
I guess the information about the sizes is in this chapter: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p3179r1.html#range_as_output.
I've checked it and It is not obvious to me whether the size selection should be done based on input+output ranges or only on input ranges. Considering all (intput+output) ranges looks reasonable, though.